### PR TITLE
upgrade: upgrade etcd before scheduler shart

### LINKED ISSUE #1644: Source unbound after upgrade
## Bug Report

Please answer these questions before submitting your issue. Thanks!

1. What did you do? If possible, provide a recipe for reproducing the error.
cluster: 1 master, 2 workers, 2 sources
upgrade from v2.0.x to nightly
2. What did you expect to see?
upgrade success
3. What did you see instead?
list-member --worker
```
"worker": {
                "msg": "",
                "workers": [
                    {
                        "name": "dm-172.28.0.111-8262",
                        "addr": "172.28.0.111:8262",
                        "stage": "bound",
                        "source": "mysql-replica-02"
                    },
                    {
                        "name": "dm-172.28.0.112-8262",
                        "addr": "172.28.0.112:8262",
                        "stage": "free",
                        "source": ""
                    }
                ]
            }
```
4. Versions of the cluster

    - DM version (run `dmctl -V` or `dm-worker -V` or `dm-master -V`):

        ```
       nightly
        ```

### PR DESCRIPTION
**Pull request by GMHDBJD**
### What problem does this PR solve? 
close #1644 

### What is changed and how it works?
upgrade etcd before scheduler shart

### Check List 

Tests 

 - github action

### CODE REVIEW

**Comment by GMHDBJD** [comment]
test upgrade-via-tiup failed because it uses nighlty version, #1635 patch with local build version and succeeds. PTAL

File: dm/master/bootstrap.go [hunk 1]
```diff
@@ -70,6 +70,20 @@ func (s *Server) bootstrap(ctx context.Context) error {
...
 }
 
+func (s *Server) bootstrapBeforeSchedulerStart(ctx context.Context) error {
+	log.L().Info("bootstrap before scheduler start")
+	// no need for v1.0.x
+	if s.cfg.V1SourcesPath != "" {
+		return nil
+	}
+
+	err := upgrade.TryUpgradeBeforeSchedulerStart(ctx, s.etcdClient)
```
Comment (lance6716) [inline dm/master/bootstrap.go:80]:
```suggestion
	return upgrade.TryUpgradeBeforeSchedulerStart(ctx, s.etcdClient)
```

Comment (lance6716) [reply dm/master/bootstrap.go:80]: seems our lint is broken 🤔

File: pkg/upgrade/upgrade.go [hunk 2]
```diff
@@ -204,7 +243,16 @@ func upgradeToVer3(cli *clientv3.Client, uctx Context) error {
 
 	ops := make([]clientv3.Op, 0, len(etcdKeyUpgrades))
 	for _, pair := range etcdKeyUpgrades {
-		resp, err := cli.Get(uctx.Context, pair.old.Path(), clientv3.WithPrefix())
+		resp, err := cli.Get(ctx, pair.new.Path(), clientv3.WithPrefix())
+		if err != nil {
+			return err
+		}
+		if len(resp.Kvs) > 0 {
```
Comment (lance6716) [inline pkg/upgrade/upgrade.go:250]: what's the benefits of we changing the upgrading detections? If user removes source configs after scheduler.Start() and DM-master failed to save current version in etcd (may caused by we write a buggy upgradeToVerX in future), the new leader will add source configs again

**Review by lance6716 — COMMENTED** [review]
rest lgtm

File: dm/master/bootstrap.go (hunk 1 above)
Comment (GMHDBJD) [reply dm/master/bootstrap.go:80]: Seems revive can only detect err declare in if
```golang
if err:=...;err!=nil{
    return err
}
return nil
```

**Comment by GMHDBJD** [comment]
merge in #1635

**Pull request closed without merging**